File: src/081.py
# Classic DP problem
import csv
N = 80
with open("p081_matrix.txt", 'r') as f:
    matrix = [list(map(int, row)) for row in csv.reader(f)]

path = [[0] * N for _ in range(N)]
for i in range(N):
    for j in range(N):
        if (i, j) == (0, 0): path[i][j] = matrix[i][j]
        elif i == 0:         path[i][j] = matrix[i][j] + path[i][j-1]
        elif j == 0:         path[i][j] = matrix[i][j] + path[i-1][j]
        else:                path[i][j] = matrix[i][j] + \
                                          min(path[i][j-1], path[i-1][j])

print(path[N-1][N-1])

File: src/081.c
/* Exercise in reading files and working with arrays */
#include <stdio.h>
#include <sys/param.h>

#define N 80 

int matrix[N][N];
int path[N][N];

int main() 
{
    FILE *fp = fopen("p081_matrix.txt", "r");
    
    for (int i=0; i<N; ++i) 
    {
        for (int j=0; j<N; ++j) 
        {
            fscanf(fp, "%d,", &matrix[i][j]); // ignore return value
            path[i][j] = matrix[i][j];
        }
    }

    for (int i=0; i<N; ++i)
    {
        for (int j=0; j<N; ++j)
        {
            if (i == 0 && j == 0) continue;
            else if (i == 0) path[i][j] += path[i][j-1];
            else if (j == 0) path[i][j] += path[i-1][j];
            else             path[i][j] += MIN(path[i][j-1], path[i-1][j]);

        }
    }
    printf("%d\n", path[N-1][N-1]);
}
